java-components/cache/src/main/java/com/redhat/hacbs/artifactcache/services/client/ociregistry/OCIRegistryRepositoryClient.java

```diff
@@ -0,0 +1,229 @@
+        CredentialRetrieverFactory credentialRetrieverFactory = CredentialRetrieverFactory.forImage(imageReference,
+                (s) -> Log.info(s.getMessage()));
+
+        RegistryClient.Factory factory = RegistryClient.factory(new EventHandlers.Builder().build(), registry,
+                owner + File.separator + group,
+                new FailoverHttpClient(enableHttpAndInsecureFailover, false, s -> Log.info(s.getMessage())));
+
+        try {
+            Optional<Credential> credential = credentialRetrieverFactory.dockerConfig().retrieve();
+            if (credential.isPresent()) {
+                factory.setCredential(credential.get());
+            } else {
+                Log.warn("No credential found for " + registry + ", proceeding without any");
+            }
+        } catch (CredentialRetrievalException cre) {
+            throw new RuntimeException(cre);
+        }
+
+        RegistryClient registryClient = factory.newRegistryClient();
+
+        return registryClient;
+    }
+
+    private Path getLocalCachePath(RegistryClient registryClient, ManifestTemplate manifest, String digestHash)
+            throws IOException {
+        if (imageCache.containsKey(digestHash)) {
+            return pullStale(registryClient, manifest, digestHash);
+        } else {
+            return pullFreshAndCache(registryClient, manifest, digestHash);
+        }
+    }
+
+    private Path pullStale(RegistryClient registryClient, ManifestTemplate manifest, String digestHash) throws IOException {
+        Path path = imageCache.get(digestHash);
+        if (Files.exists(path)) {
+            return path;
+        } else {
+            return pullFreshAndCache(registryClient, manifest, digestHash);
+        }
+
+    }
+
+    private Path pullFreshAndCache(RegistryClient registryClient, ManifestTemplate manifest, String digestHash)
+            throws IOException {
+
+        String manifestMediaType = manifest.getManifestMediaType();
+
+        if (OCI_MEDIA_TYPE.equalsIgnoreCase(manifestMediaType)) {
+            List<BuildableManifestTemplate.ContentDescriptorTemplate> layers = ((OciManifestTemplate) manifest).getLayers();
+
+            // Layer 2 is artifacts
+            BuildableManifestTemplate.ContentDescriptorTemplate artifactsLayer = layers.get(2);
+
+            Blob blob = registryClient.pullBlob(artifactsLayer.getDigest(), s -> {
+            }, s -> {
+            });
+
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            blob.writeTo(out);
+
+            // Extract to temp (TODO: Configure this ?)
+            String osTmpDir = System.getProperty(SYS_PROP_TEMP_DIR);
+            Path tempFullPath = Paths.get(osTmpDir, HACBS);
+            Path outputPath = Files.createDirectories(tempFullPath);
+            extractTarArchive(out.toByteArray(), outputPath.toString());
+            Path repoRoot = Paths.get(outputPath.toString(), ARTIFACTS);
```

Comment:
Shouldn't this have the hash in the path? Otherwise they are all getting mixed into the same directory?

Reply:
Not really as the version is in the path, but now the hash will be in the path anyway

Reply:
What do you mean? The path is just  osTmpDir + HACBS ?

Reply:
Yea but then follow it a mvn like structure that contains the version.

Reply:
This does allow images to potentially overwrite each other though. These really need to be in a directory that matches the hash. You can also then use the existence of this directory to know if the layer is present or not.

Reply:
Yes, that is what I am doing not. So removing the map that holds the cache and just look at local cache